File: Math/C++/10.Greatest Common Divisor of Strings.cpp
/*Algorithm: 
Considering smaller string as str1, we will consider all substrings of str1 which can be GCD ,starting from maximum length.
If the length of any substring of str1 is multiple of length of str1 and str2 and divide both str1 and str2, then that substring
is GCD of two given strings.
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string gcdOfStrings(string str1, string str2) 
    {
        if(str1.length()>str2.length()) //considering str1 as smaller string
            swap(str1,str2);
        
        for(int n=str1.length();n>=1;n--)//considering all possible substring
        {
            if(str1.length()%n!=0) //size of substring is not multiple of str1
                continue;
            
            if(str2.length()%n!=0)//size of substring is not multiple of str1
                continue;
            
            string s=str1.substr(0,n); //possible substring
            
            int f=0;
            
            for(int i=0;i<str1.length();i+=n)
            {
                if(s!=str1.substr(i,n)) // substring s is not diving str1
                {
                    f=1;
                    break;
                }
            }
            
            if(f==1)
                continue;
            
            f=0;
                
            for(int i=0;i<str2.length();i+=n)
            {
                if(s!=str2.substr(i,n)) // substring s is not diving str2
                {
                    f=1;
                    break;
                }
            }
            
            if(f==0)
                return s; //substring found
            
            
            
        }
        return "";
        
    }
};

int main()
{
    string s,t;
    cin>>s>>t;

    Solution obj;
    string ans=obj.gcdOfStrings(s,t);
    
    cout<<ans<<endl;


}

//Time Complexity: O(n*n)
//Space Complexity: O(1)

/*
Input: str1 = "ABABAB", str2 = "ABAB"
Output: "AB"
Explanation : AB is the longest positive string which can divide str1 and str2.

Input: str1 = "LEET", str2 = "CODE"
Output: ""
Explanation : There is no such string which can divide both str1 and str2

*/

File: Math/Java/10_Greatest Common Divisor of Strings.java
/*
Given two strings str1 and str2, return the largest string x such that x divides both str1 and str2.

In strings ‘str1’ and ‘str2’, we say “str2 divides str1” if and only if str1 = concatenation of str2 more than 1 times. Here we have to find the largest string which divides both str1 and str2. 

Input: str1 = "ABCABC", str2 = "ABC"
Output: "ABC"

Input: str1 = "ABABAB", str2 = "ABAB"
Output: "AB"
*/

//The code below implements this approach

import java.util.Scanner;

public class GreatestCommonDivisor {

    static Scanner sc = null;

    public String gcdOfStrings(String str1, String str2) {

        // Checking if the two strings are already equal
        while (!str2.isEmpty()) {
            if (str1.equals(str2)) {
                return str1;
            }

            // swap to make str1 longer than str2
            if (str1.length() < str2.length()) {
                String tmp = str1;
                str1 = str2;
                str2 = tmp;
            }

            // checking whether the substring of str1 (length of which is equal to the
            // length of str2) is equal to str2
            if (str1.substring(0, str2.length()).equals(str2)) {
                String r = str1.substring(str2.length());

                str1 = str2;
                str2 = r;

            } else {
                return "false";
            }
        }

        return str1;

    }

    public static void main(String[] args) {
        sc = new Scanner(System.in);
        System.out.println("Enter the first string : ");
        String str1 = sc.nextLine();
        System.out.println("Enter the second string : ");
        String str2 = sc.nextLine();
        GreatestCommonDivisor obj = new GreatestCommonDivisor();
        String res = obj.gcdOfStrings(str1, str2);
        if (res.equals("false")) {
            System.out.println("No Greatest Common Divisor possible!");
        } else
            System.out.println("The Greatest Common Divisor is : " + obj.gcdOfStrings(str1, str2));
    }

}

/*
Sample Input:
Enter the first string : 
XYZXYZ
Enter the second string : 
XYZ
Sample Output:
The Greatest Common Divisor is : XYZ

Sample Input:
Enter the first string : 
ABCABS
Enter the second string : 
ABC
Sample Output:
No Greatest Common Divisor possible!
*/
